criu/files-reg.c

```diff
@@ -1380,12 +1381,299 @@ static bool should_check_size(int flags)
+	Elf64_Nhdr *note_header, *note_header_end;
+
+	/* 
+	 * If the file doesn't have atleast 1 program header entry, it definitely can't
+	 * have a build-id.
+	 */
+	if (!file_header->e_phnum)
+		goto err;
+
+	/* 
+	 * If the build-id exists, then it will most likely be present in the
+	 * beginning of the file. Therefore at most only the first 1 MB of the 
+	 * file was mapped.
+	 */
+	file_header_end = min_t(size_t, fd_status->st_size, 1048576) + (size_t) file_header;
+
+	program_header = (Elf64_Phdr *) (file_header->e_phoff + (char *) file_header);
+	if (program_header <= (Elf64_Phdr *) file_header)
+		goto err;
+
+	program_header_end = (Elf64_Phdr *) (file_header_end - sizeof(Elf64_Phdr));
+	num_iterations = file_header->e_phnum + 1;
+
+	/* 
+	 * If the file has a build-id, it will be in the PT_NOTE program header 
+	 * entry AKA the note sections.
+	 */
+	while (num_iterations-- && program_header <= program_header_end &&
+			program_header->p_type != PT_NOTE)
+		program_header++;
+
+	if (!num_iterations || program_header >= program_header_end)
+		goto err;
+
+	note_header = (Elf64_Nhdr *) (program_header->p_offset + (char *) file_header);
+	if (note_header <= (Elf64_Nhdr *) file_header)
+		goto err;
+
+	note_header_end = (Elf64_Nhdr *) (file_header_end - sizeof(Elf64_Nhdr));
+	/* Arbitrary large number to prevent unnecessarily searching the entire file */
+	num_iterations = 500;
+
+	/* The note type for the build-id is NT_GNU_BUILD_ID. */
+	while (num_iterations-- && note_header <= note_header_end &&
+			note_header->n_type != NT_GNU_BUILD_ID)
+		note_header = (Elf64_Nhdr *) ((char *) note_header + sizeof(Elf64_Nhdr) +
+						note_header->n_namesz + note_header->n_descsz);
+
+	if (!num_iterations || note_header >= note_header_end)
+		goto err;
+
+	/* 
+	 * If the size of the notes description is too large or is invalid
+	 * then the build-id could not be obtained.
+	 */
+	if (note_header->n_descsz <= 0 || note_header->n_descsz > 512)
+		goto err;
+
+	size = note_header->n_descsz;
+	note_header = (Elf64_Nhdr *) ((char *) note_header + sizeof(Elf64_Nhdr) +
+					note_header->n_namesz);
+	note_header_end = (Elf64_Nhdr *) (file_header_end - size);
+	if (note_header <= (Elf64_Nhdr *) file_header || note_header > note_header_end)
+		goto err;
+
+	*build_id = (unsigned char *) xmalloc(size);
+	if (!*build_id)
+		goto err;
+
+	memcpy(*build_id, (void *) note_header, size);
+
+	munmap(file_header, min_t(size_t, fd_status->st_size, 1048576));
+	return size;
+	
+err:
+	munmap(file_header, min_t(size_t, fd_status->st_size, 1048576));
+	return -1;
+}
+
+/*
+ * Finds the build-id of the file by checking if the file is an ELF file
+ * and then calling either the 32-bit or the 64-bit function as necessary.
+ * Returns the number of bytes of the build-id if it could be
+ * obtained, else -1.
+ */
+static int get_build_id(const int fd, const struct stat *fd_status,
+				unsigned char **build_id)
+{
+	char *start_addr, buf[5];
+
+	if (read(fd, buf, 5) != 5)
```

Comment:
read(fd, buf, SELFMAG) != SELFMAG

Reply:
But I will also need the 5th bit (SELFMAG + 1) since I need to know whether the file is 32-bit or 64-bit right?